tests/plugins/colordialog/colordialog.js

```diff
@@ -66,8 +66,49 @@
 		// (#2639)
 		'test background color of mixed selection': function() {
 			openDialogManually( this.editor, '', '<h1>[<span style="background-color:#0000ff">Foo</span>bar]</h1>', 'BGColor' );
-		}
+		},
+
+		// (#4351)
+		'test converting RGBA to hex color': function() {
+			// IE8 doesn't set unsupported values.
+			if ( CKEDITOR.env.ie && CKEDITOR.env.version < 9 ) {
+				assert.ignore();
+			}
+
+			assertSettingAndGettingColor( this.editor, {
+				inputColor: 'rgba(100,200,50,.4)',
+				expectedColor: '#c1e9ad',
+				button: 'TextColor'
+			} );
+		},
 
+		// (#4351)
+		'test converting HSL to hex color': function() {
+			// IE8 doesn't set unsupported values.
+			if ( CKEDITOR.env.ie && CKEDITOR.env.version < 9 ) {
+				assert.ignore();
+			}
+
+			assertSettingAndGettingColor( this.editor, {
+				inputColor: 'hsl(150,50%,52%)',
+				expectedColor: '#47c285',
+				button: 'TextColor'
+			} );
+		},
+
+		// (#4351)
+		'test converting HSLA to hex color': function() {
+			// IE8 doesn't set unsupported values.
+			if ( CKEDITOR.env.ie && CKEDITOR.env.version < 9 ) {
+				assert.ignore();
+			}
+
+			assertSettingAndGettingColor( this.editor, {
+				inputColor: 'hsla(150,50%,52%,0.2)',
+				expectedColor: '#daf3e7',
+				button: 'TextColor'
+			} );
+		}
```

Comment:
I would also test the transparent color code case described by you in the comment:
```js
// Browsers tend to represent transparent color as rgba(0, 0, 0, 0), so we need to filter out this value.
```